File: backend/nlp/nlu.js
/**
 * =========================================================
 * ORBITAL NLU ENGINE
 * =========================================================
 *
 * Converts tokenized human language into structured meaning.
 *
 * Example:
 *
 * "Show me the highest risk debris near Earth"
 *
 * becomes:
 *
 * {
 *   intent: "highest_risk",
 *   entities: {
 *     object_type: "debris",
 *     location: "Earth"
 *   }
 * }
 *
 * This first version uses simple rules.
 */

const { tokenize } = require("./tokenizer");
const { normalizeTokens } = require("./normalizer");

/*
=========================================================
INTENT DETECTION
=========================================================
*/

function detectIntent(tokens) {

  /*
  -------------------------------------------------------
  HIGHEST RISK
  -------------------------------------------------------
  */

  if (
    tokens.includes("highest") &&
    (
      tokens.includes("risk") ||
      tokens.includes("dangerous")
    )
  ) {
    return "highest_risk";
  }


  /*
  -------------------------------------------------------
  COLLISION LOOKUP
  -------------------------------------------------------
  */

  if (
    tokens.includes("collision") ||
    tokens.includes("collisions")
  ) {
    return "collision_lookup";
  }


  /*
  -------------------------------------------------------
  TRACK OBJECT
  -------------------------------------------------------
  */

  if (
    tokens.includes("track") ||
    tokens.includes("tracking")
  ) {
    return "track_object";
  }


  /*
  -------------------------------------------------------
  EXPLAIN SGP4
  -------------------------------------------------------
  */

  if (
    tokens.includes("sgp4")
  ) {
    return "explain_sgp4";
  }


  /*
  -------------------------------------------------------
  UNKNOWN
  -------------------------------------------------------
  */

  return "unknown";
}


/*
=========================================================
ENTITY EXTRACTION
=========================================================
*/

function extractEntities(tokens) {

  const entities = {};


  /*
  -------------------------------------------------------
  OBJECT TYPE
  -------------------------------------------------------
  */

  if (
    tokens.includes("debris")
  ) {
    entities.object_type = "debris";

  } else if (
    tokens.includes("satellite")
  ) {
    entities.object_type = "satellite";
  }


  /*
  -------------------------------------------------------
  LOCATION
  -------------------------------------------------------
  */

  if (
    tokens.includes("earth")
  ) {
    entities.location = "Earth";
  }


  /*
  -------------------------------------------------------
  NORAD ID
  -------------------------------------------------------
  */

  const noradIndex =
    tokens.indexOf("norad");

  if (
    noradIndex !== -1 &&
    tokens[noradIndex + 1]
  ) {

    const candidate =
      tokens[noradIndex + 1];

    if (
      /^\d+$/.test(candidate)
    ) {
      entities.norad_id =
        candidate;
    }
  }


  /*
  -------------------------------------------------------
  RETURN
  -------------------------------------------------------
  */

  return entities;
}


/*
=========================================================
MAIN NLU FUNCTION
=========================================================
*/

function understand(text) {

  const rawTokens =
  tokenize(text);

const tokens =
  normalizeTokens(rawTokens);

  const intent =
    detectIntent(tokens);

  const entities =
    extractEntities(tokens);

  return {
  text,
  rawTokens,
  tokens,
  intent,
  entities,
};
}


/*
=========================================================
EXPORT
=========================================================
*/

module.exports = {
  understand,
  detectIntent,
  extractEntities,
};

File: ai/inference/nlu.py
import re

from ai.inference.intent_predictor import predict_intent


def extract_entities(text):
    """
    Extract important orbital entities from user text.
    """

    entities = {}

    text_lower = text.lower()

    # -------------------------------------------------
    # NORAD ID
    # -------------------------------------------------

    norad_match = re.search(
        r"\b(?:norad(?:\s+id)?|object|satellite)\s*#?\s*(\d{4,6})\b",
        text_lower
    )

    if norad_match:
        entities["norad_id"] = norad_match.group(1)

    # -------------------------------------------------
    # Object type
    # -------------------------------------------------

    if "debris" in text_lower:
        entities["object_type"] = "debris"

    elif "satellite" in text_lower:
        entities["object_type"] = "satellite"

    # -------------------------------------------------
    # Location
    # -------------------------------------------------

    if "earth" in text_lower:
        entities["location"] = "Earth"

    return entities


def understand(text):
    """
    Combine ML intent classification
    with entity extraction.
    """

    intent_result = predict_intent(text)

    entities = extract_entities(text)

    return {
        "text": text,
        "intent": {
            "name": intent_result["intent"],
            "confidence": intent_result["confidence"],
        },
        "entities": entities,
    }